### ISSUE TITLE
problem installing Jupyterhub on VPS (Debian 9)

### ISSUE THREAD

**Issue by msmarchena** [jupyterhub/jupyterhub#2250]
I have installed Jupyterhub on my VPS (Debian 9) using nginx as proxy. The problem is that when I use my domain I have 502 Bad Gateway error. I have tried all possible solutions but nothing works.

Bellow my configuration and some outputs

    (jupyterhub-tutorial)  .. $>jupyterhub
    [I 2018-10-18 12:15:32.654 JupyterHub app:622] Loading cookie_secret from /srv/jupyterhub/jupyterhub-pydata-2016/jupyterhub_cookie_secret
    [W 2018-10-18 12:15:32.732 JupyterHub app:304] 
    Generating CONFIGPROXY_AUTH_TOKEN. Restarting the Hub will require restarting the proxy.
    Set CONFIGPROXY_AUTH_TOKEN env or JupyterHub.proxy_auth_token config to avoid this message.
    
    [W 2018-10-18 12:15:32.738 JupyterHub app:757] No admin users, admin interface will be unavailable.
    [W 2018-10-18 12:15:32.739 JupyterHub app:758] Add any administrative users to `c.Authenticator.admin_users` in config.
    [I 2018-10-18 12:15:32.739 JupyterHub app:785] Not using whitelist. Any authenticated user will be allowed.
    [I 2018-10-18 12:15:32.756 JupyterHub app:1231] Hub API listening on http://127.0.0.1:8081/hub/
    [I 2018-10-18 12:15:32.760 JupyterHub app:968] Starting proxy @ http://127.0.0.1:8000/
    12:15:33.296 - info: [ConfigProxy] Adding route / -> http://127.0.0.1:8081
    12:15:33.313 - info: [ConfigProxy] Proxying https://127.0.0.1:8000 to http://127.0.0.1:8081
    12:15:33.314 - info: [ConfigProxy] Proxy API at http://127.0.0.1:8001/api/routes
    [I 2018-10-18 12:15:33.374 JupyterHub app:1254] JupyterHub is now running at http://127.0.0.1:8000/

My jupyterhub_config.py only have these lines activated

    c.JupyterHub.ssl_key = '/etc/letsencrypt/live/mysite.com/privkey.pem'
    c.JupyterHub.ssl_cert = '/etc/letsencrypt/live/mysite.com/fullchain.pem'
    c.JupyterHub.ip = '127.0.0.1'
    c.JupyterHub.base_url = '/'

and my nginx.conf is

    user www-data;
    worker_processes 4;
    pid /run/nginx.pid;

    events {
      worker_connections 1024;
    }

    http {

            include /etc/nginx/mime.types;
            default_type application/octet-stream;

        server {
            listen 80;
            server_name mysite.com;
            rewrite        ^ https://$host$request_uri? permanent;
        }

        server {
            listen 443;
            client_max_body_size 50M;

            server_name mysite.com;

            ssl on;
            ssl_certificate /etc/letsencrypt/live/mysite.com/fullchain.pem;
            ssl_certificate_key /etc/letsencrypt/live/mysite.com/privkey.pem;

            ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
            ssl_prefer_server_ciphers on;
            ssl_dhparam /etc/ssl/certs/dhparam.pem;
            ssl_ciphers 'ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA';
            ssl_session_timeout 1d;
            ssl_session_cache shared:SSL:50m;
            ssl_stapling on;
            ssl_stapling_verify on;
            add_header Strict-Transport-Security max-age=15768000;

            # Expose logs to "docker logs".
            # See https://github.com/nginxinc/docker-nginx/blob/master/Dockerfile#L12-L14
            access_log /var/log/nginx/access.log;
            error_log /var/log/nginx/error.log;

            #location ~ /(user-[a-zA-Z0-9]*)/static(.*) {
            #    alias /usr/local/lib/python3.4/dist-packages/notebook/static/$2;
           #}

             location / {
                proxy_pass http://127.0.0.1:8000;

                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header Host $host;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

                proxy_set_header X-NginX-Proxy true;
            }

            location ~* /(user/[^/]*)/(api/kernels/[^/]+/channels|terminals/websocket)/? {
                proxy_pass http://127.0.0.1:8000;

                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header Host $host;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

                proxy_set_header X-NginX-Proxy true;

                # WebSocket support
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection "upgrade";
                proxy_read_timeout 86400;
                proxy_set_header Origin "";
            }
        }

    }

some system outputs

    $>sudo ufw status

    To                         Action      From
    --                         ------      ----
    22                         ALLOW       Anywhere                  
    80                         ALLOW       Anywhere                  
    443                        ALLOW       Anywhere                  
    8000                       ALLOW       Anywhere                  
    Nginx HTTP                 ALLOW       Anywhere                  
    Nginx HTTPS                ALLOW       Anywhere                  
    Nginx Full                 ALLOW       Anywhere                  
    8001                       ALLOW       Anywhere                  
    22 (v6)                    ALLOW       Anywhere (v6)             
    80 (v6)                    ALLOW       Anywhere (v6)             
    443 (v6)                   ALLOW       Anywhere (v6)             
    8000 (v6)                  ALLOW       Anywhere (v6)             
    Nginx HTTP (v6)            ALLOW       Anywhere (v6)             
    Nginx HTTPS (v6)           ALLOW       Anywhere (v6)             
    Nginx Full (v6)            ALLOW       Anywhere (v6)             
    8001 (v6)                  ALLOW       Anywhere (v6)     

    $> netstat -an | grep "LISTEN "

    tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN     
    tcp        0      0 127.0.0.1:8081          0.0.0.0:*               LISTEN     
    tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN     
    tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN     
    tcp        0      0 127.0.0.1:8000          0.0.0.0:*               LISTEN     
    tcp        0      0 127.0.0.1:8001          0.0.0.0:*               LISTEN     
    tcp6       0      0 :::22                   :::*                    LISTEN  

    $> wget localhost:8000
    --2018-10-18 11:32:37--  http://localhost:8000/
    Resolving localhost (localhost)... 127.0.0.1
    Connecting to localhost (localhost)|127.0.0.1|:8000... connected.
    HTTP request sent, awaiting response... No data received.
    Retrying.

    $> nmap -A -T4 mysite.com

    Starting Nmap 7.40 ( https://nmap.org ) at 2018-10-18 11:33 CEST
    Nmap scan report for mysite.com (ip)
    Host is up (0.00012s latency).
    rDNS record for ip
    Not shown: 997 closed ports
    PORT    STATE SERVICE  VERSION
    22/tcp  open  ssh      OpenSSH 7.4p1 Debian 10+deb9u4 (protocol 2.0)
    |   ssh-hostkey: 
     ....
    80/tcp  open  http     nginx 1.10.3
    |_http-server-header: nginx/1.10.3
    |_http-title: Did not follow redirect to https://mysite.com/
    443/tcp open  ssl/http nginx 1.10.3
    |_http-server-header: nginx/1.10.3
    |_http-title: 502 Bad Gateway
    | ssl-cert: Subject: commonName=mysite.com
    | Subject Alternative Name: DNS:mysite.com
    | Not valid before: 2018-10-09T12:01:25
    |_Not valid after:  2019-01-07T12:01:25
    |_ssl-date: TLS randomness does not represent time
    | tls-nextprotoneg: 
    |_  http/1.1
    Service Info: OS: Linux; CPE: cpe:/o:linux:linux_kernel

I don't have any idea what I can do to solve the problem. Any help will be much appreciated.

**Comment by BertR** [comment]
Hi @msmarchena it could have something to do with where you are teminating SSL.
If Jupyterhub is running HTTPS on port 8000 and HTTP on port 8081,
maybe you should set NGINX so the proxy_pass is going to the HTTP port and not HTTPS.

Since Jupyterhub is running locally (127.0.0.1) perhaps it is better not to configure any HTTPS
for Juptyterhub at all and setup your SSL certificate on NGINX instead.

**Comment by msmarchena** [comment]
Hi @BertR  I have solved the problem. 
In jupyterhub_config.py, I removed the SSL configuration and replaced `c.JupyterHub.ip='127.0.0.1' ` to `c.JupyterHub.ip.='ip from my VPS' `
In the nginx.conf, I changed 'proxy_pass http://127.0.0.1:8000;' to 'proxy_pass http://mysite.com:8000;' After doing these modifications I was able to see the main login page but I had

     500 : Internal Server Error
     Spawner failed to start [status=1]

when trying to log using my user server. I solved the issue after running 

    pip install --upgrade jupyterhub

Now it is working. Thanks for your help!
